tests/cpp/test_multidevice_tutorial.cpp

```diff
@@ -0,0 +1,381 @@
+  // or the mesh with device "1" only
+  DeviceMesh mesh_one({1});
+  // We can also consider the mesh of device indices "0" "2"
+  DeviceMesh mesh_zero_and_two({0, 2});
+  // Or the mesh containing all available devices in the world communicator
+  std::vector<int64_t> all_devices(communicator->size());
+  std::iota(
+      all_devices.begin(),
+      all_devices.end(),
+      0); // all_devices = [0,1,..., communicator->size()-1]
+  DeviceMesh mesh_full(all_devices);
+  // However, it is forbidden to define a mesh with duplicates indices:
+  EXPECT_ANY_THROW(DeviceMesh mesh_with_duplicates({1, 1}));
+
+  // Device meshes are used to indicate the devices on which a Tensor is
+  // sharded or replicated. Each tensor can be associated with a DeviceMesh. For
+  // example,
+  tv0->setDeviceMesh(mesh_zero);
+  // means that tv0 is only present on device "0".
+  if (verbose_print) {
+    std::cout << tv0->toString() << std::endl;
+  }
+  // Alternatively:
+  tv0->setDeviceMesh(mesh_full);
+  // means that tv0 is replicated on all the devices.
+  if (verbose_print) {
+    std::cout << tv0->toString() << std::endl;
+  }
+
+  // Let us exercise this notion by examining two simple situations where we set
+  // the device mesh. In those simple examples, no network communication will be
+  // needed at all. We will simply see that only through setting the device mesh
+  // we can, e.g., replicate computation on all devices or select a subset of
+  // devices.
+
+  // FULL REPLICATION ON ALL DEVICES
+  // Let us set tv0 and tv1 to be replicated on all the devices:
+  tv0->setDeviceMesh(mesh_full);
+  tv1->setDeviceMesh(mesh_full);
+
+  // RUNTIME
+  // Set up the input
+  constexpr int64_t tensor_size = 128;
+  const c10::TensorOptions tensor_options =
+      at::TensorOptions().device(communicator->device()).dtype(at::kFloat);
+  // each rank allocate a tensor a on different device
+  at::Tensor input = at::randn({tensor_size}, tensor_options);
+  {
+    // EXECUTION
+    // This class is responsible for managing a single device (given by
+    // communicator->deviceId()) in a SPMD multidevice program. Recall that each
+    // rank manages one and only one GPU.
+    MultiDeviceExecutor multidevice_executor(
+        std::make_unique<Fusion>(fusion), *communicator);
+    if (verbose_print) {
+      fusion.printMath();
+      fusion.printKernel();
+      multidevice_executor.print();
+    }
+
+    // The same computation is replicated on all the devices
+    at::Tensor output = multidevice_executor.runWithInput({input}).at(0);
+
+    // VALIDATION
+    // Each device produces the full output
+    EXPECT_TRUE(output.equal(input));
+  }
+
+  // SINGLE DEVICE EXECUTION
+  // This indicates that tv0 and tv1 are present on device 0 only
+  tv0->setDeviceMesh(mesh_zero);
+  tv1->setDeviceMesh(mesh_zero);
+  // RUNTIME
+  {
+    // EXECUTION
+    MultiDeviceExecutor multidevice_executor(
+        std::make_unique<Fusion>(fusion), *communicator);
```

Comment:
Make `fusion` a unique_ptr and do
```suggestion
        std::move(fusion), *communicator);
```
This saves a fusion copy.

Reply:
Right, well spotted! However, this can't work as is because `fusion` is of non-pointer type `Fusion`. So I'll leave it as is for now, despite the unnecessary fusion copy. ok? lmk

Reply:
That's why I said make fusion an uniqueptr. Would that not work?

Reply:
https://github.com/NVIDIA/Fuser/pull/2733